ops/cli.py

```diff
@@ -1,351 +1,343 @@
+        return val
+
+    @staticmethod
+    def __read_version() -> str:
+        with open("version") as version_file:
+            return version_file.read().strip()
+
+    def __find_cloudsql_name(self) -> str:
+        """Use labels to identify the Cloud SQL name."""
+        instances = json.loads(shell(f"gcloud --project {self.project} --format=json "
+                                     "sql instances list "
+                                     f"--filter='labels.app_name=wfl AND labels.instance_id={self.instance_id}'",
+                                     quiet=True))
+        if len(instances) == 1:
+            return instances[0]["name"]
+        else:
+            error("No CloudSQL instance could be inferred")
+            info("    Couldn't find a single match for "
+                 f"`--filter='labels.app_name=wfl AND labels.instance_id={self.instance_id}'`")
+            info(f"    Available instances in {self.project}:", plain=True)
+            info(shell(f"gcloud --project {self.project} sql instances list", quiet=True), plain=True)
+            return None if self.force else exit(1)
+
+    def __find_cluster_name(self) -> str:
+        """Look at the Cloud SQL labels to find the cluster name."""
+        instances = json.loads(shell(f"gcloud --project {self.project} --format=json "
+                                     "sql instances list "
+                                     f"--filter='labels.app_name=wfl AND labels.instance_id={self.instance_id}'",
+                                     quiet=True))
+        if len(instances) == 1:
+            return instances[0]["labels"]["app_cluster"]
+        else:
+            error("No cluster name could be inferred")
+            info("    Couldn't find a single match for "
+                 f"`--filter='labels.app_name=wfl AND labels.instance_id={self.instance_id}'`")
+            info("    The `app_cluster` label is used to find cluster name, see wfl-instance documentation.")
+            info(f"    Available clusters in {self.project}:", plain=True)
+            info(shell(f"gcloud --project {self.project} container clusters list", quiet=True), plain=True)
+            return None if self.force else exit(1)
+
+    def __find_cluster_zone(self) -> str:
+        """Look in the cluster list for the cluster's zone."""
+        clusters = json.loads(shell(f"gcloud --project {self.project} --format=json "
+                                    "container clusters list "
+                                    f"--filter='name={self.cluster_name}'"))
+        if len(clusters) == 1:
+            return clusters[0]["zone"]
+        else:
+            error(f"No cluster zone could be found for cluster named {self.cluster_name}")
+            info("    Couldn't find a single match for "
+                 f"`--filter='name={self.cluster_name}'`")
+            return None if self.force else exit(1)
+
+    def validate(self):
```

Comment:
I don't understand why you've got a separate validation step.
Error when you encounter errors - don't save them for later and duplicate logic. 

In the olden days of mainframes and time sharing it was useful for your compiler to keep going and catch as many errors as possible. Nowadays, this doesn't actually help because later errors often depend on earlier ones. TLDR: don't accumulate errors.

Reply:
Validating as a secondary step is a simple way to always validate even if you supply overrides for certain arguments. Making it one step wouldn't actually abstract away much code because all the same things still need to happen--whether it is "specific or calculate default, then validate" (as here) or "specific and validate or calculate default" (as you suggest) there isn't much of a change in the complexity. I think the order I chose gives a good demarcation for splitting it up.

The reason for having a `force` flag and allowing all these to be bypassed is so that the `connect` command can be used even if some parts of the wfl-instance aren't online or set up correctly. @rexwangcc said he didn't see a problem with running the same validation everywhere so long as it could be skipped if need be

Reply:
The `force` flag is weird. You've added nulls when they're not necessary. You've got functions like `__find_cluster_name` and `__cluster_name_exists` where you're validating twice. It makes no sense.

Reply:
The former function is used as a default while the latter is meant to check supplied values. I agree that there's a common case where we'll validate twice--we generate the default and then pass it through the validator. Alternatively we can validate only passed inputs, and that's a totally fair--I mentioned above that taking that route wouldn't actually reduce much code since we still need both pieces of functionality.

Reply:
Solved in the new version. Individual functions handle inference and validation so that validation only occurs on manually customized arguments. Any errors cause immediate exits.